paasta_tools/kubernetes_tools.py

```diff
@@ -0,0 +1,581 @@
+        )
+
+    def __repr__(self) -> str:
+        return "KubernetesDeploymentConfig({!r}, {!r}, {!r}, {!r}, {!r}, {!r})".format(
+            self.service,
+            self.cluster,
+            self.instance,
+            self.config_dict,
+            self.branch_dict,
+            self.soa_dir,
+        )
+
+    def copy(self) -> "KubernetesDeploymentConfig":
+        return self.__class__(
+            service=self.service,
+            instance=self.instance,
+            cluster=self.cluster,
+            config_dict=dict(self.config_dict),
+            branch_dict=dict(self.branch_dict) if self.branch_dict is not None else None,
+            soa_dir=self.soa_dir,
+        )
+
+    def get_bounce_method(self) -> str:
+        """Get the bounce method specified in the service's kubernetes configuration."""
+        # map existing bounce methods to k8s equivalents.
+        kube_deploy_stategy_map = {'crossover': 'RollingUpdate', 'downthenup': 'Recreate'}
+        return kube_deploy_stategy_map[self.config_dict.get('bounce_method', 'crossover')]
+
+    def get_deployment_strategy_config(self) -> Dict[str, Any]:
+        strategy_dict: Dict[str, Any] = {}
+        strategy_dict['type'] = self.get_bounce_method()
+        if self.get_bounce_method() == 'RollingUpdate':
+            # this translates bounce_margin to k8s speak maxUnavailable
+            # for now we keep max_surge 100% but we could customise later
+            strategy_dict['rollingUpdate'] = {
+                "maxSurge": '100%',
+                "maxUnavailable": "{}%".format(int((1 - self.get_bounce_margin_factor()) * 100)),
+            }
+        return strategy_dict
+
+    def get_sanitised_volume_name(self, volume_name: str) -> str:
+        """I know but we really aren't allowed many characters..."""
+        volume_name = volume_name.rstrip('/')
+        sanitised = volume_name.replace('/', 'slash-')
+        return sanitised.replace('_', '--')
+
+    def get_sidecar_containers(self, system_paasta_config: SystemPaastaConfig) -> List[Dict[str, Any]]:
+        registrations = " ".join(self.get_registrations())
+        hacheck_sidecar = {
+            "image": system_paasta_config.get_hacheck_sidecar_image_url(),
+            "lifecycle": {
+                "preStop": {
+                    "exec": {
+                        "command": [
+                            "/bin/sh",
+                            "-c",
+                            f"/usr/bin/hadown {registrations}; sleep 31",
+                        ],
+                    },
+                },
+            },
+            "name": "hacheck",
+            "env": self.get_kubernetes_environment(),
+            "ports": [
+                {
+                    "containerPort": 6666,
+                },
+            ],
+            "volumeMounts": [
+                {
+                    "mountPath": "/var/spool/hacheck",
+                    "name": self.get_sanitised_volume_name("/var/spool/hacheck"),
```

Comment:
why'd you need to mount this? because hacheck doesn't create it itself?

Reply:
huh, I can't remember why I thought this would be needed, I think you are right that it isn't needed.